File: site/netlify/functions/macros.mjs
import { getStore } from "@netlify/blobs";

export default async (req) => {
  const store = getStore("darvis-macros");

  if (req.method === "GET") {
    let macros = {};
    try {
      const data = await store.get("all", { type: "json" });
      if (data && typeof data === "object") macros = data;
    } catch {}
    return Response.json({ macros });
  }

  if (req.method === "POST") {
    const body = await req.json();

    // Full replacement from terminal sync
    if (body.replace && body.macros) {
      await store.setJSON("all", body.macros);
      return Response.json({ ok: true });
    }

    // Single macro add
    if (!body.name || !body.command) {
      return Response.json({ error: "Need name and command" }, { status: 400 });
    }

    let macros = {};
    try {
      const data = await store.get("all", { type: "json" });
      if (data) macros = data;
    } catch {}

    macros[body.name.toLowerCase()] = body.command;
    await store.setJSON("all", macros);
    return Response.json({ ok: true });
  }

  if (req.method === "DELETE") {
    const body = await req.json();
    if (!body.name) return Response.json({ error: "Need name" }, { status: 400 });

    let macros = {};
    try {
      const data = await store.get("all", { type: "json" });
      if (data) macros = data;
    } catch {}

    delete macros[body.name.toLowerCase()];
    await store.setJSON("all", macros);
    return Response.json({ ok: true });
  }

  return new Response("Method not allowed", { status: 405 });
};

export const config = { path: "/api/macros" };

File: macros.py
"""
D.A.R.V.I.S. Voice Macros — custom shortcuts that expand into actions.
Cloud-synced via Netlify Blobs. Local fallback to macros.json.
"""

import json
import urllib.request
import urllib.error
import threading
from pathlib import Path

CLOUD_URL = "https://darvis1.netlify.app/api/macros"
LOCAL_PATH = Path(__file__).parent / "macros.json"
TIMEOUT = 5


class MacroManager:
    def __init__(self):
        self._lock = threading.Lock()
        self.macros: dict[str, str] = {}
        self._load_local()

    def _load_local(self):
        if LOCAL_PATH.exists():
            try:
                self.macros = json.loads(LOCAL_PATH.read_text())
            except (json.JSONDecodeError, OSError):
                self.macros = {}

    def _save_local(self):
        LOCAL_PATH.write_text(json.dumps(self.macros, indent=2))

    def sync_from_cloud(self):
        try:
            req = urllib.request.Request(CLOUD_URL, method="GET")
            with urllib.request.urlopen(req, timeout=TIMEOUT) as resp:
                data = json.loads(resp.read().decode())
                cloud_macros = data.get("macros", {})
                if cloud_macros:
                    with self._lock:
                        self.macros.update(cloud_macros)
                        self._save_local()
        except Exception:
            pass

    def _sync_to_cloud(self):
        try:
            payload = json.dumps({"replace": True, "macros": self.macros}).encode()
            req = urllib.request.Request(
                CLOUD_URL, data=payload, method="POST",
                headers={"Content-Type": "application/json"})
            urllib.request.urlopen(req, timeout=TIMEOUT)
        except Exception:
            pass

    def add(self, name: str, command: str) -> str:
        with self._lock:
            self.macros[name.lower()] = command
            self._save_local()
        threading.Thread(target=self._sync_to_cloud, daemon=True).start()
        return f"Macro '{name}' saved: {command}"

    def remove(self, name: str) -> str:
        with self._lock:
            if name.lower() in self.macros:
                del self.macros[name.lower()]
                self._save_local()
                threading.Thread(target=self._sync_to_cloud, daemon=True).start()
                return f"Macro '{name}' removed"
            return f"Macro '{name}' not found"

    def get(self, name: str) -> str | None:
        with self._lock:
            return self.macros.get(name.lower())

    def list_all(self) -> dict[str, str]:
        with self._lock:
            return dict(self.macros)
